### PR TITLE
Use chip::ByteSpan for CHAR_STRING and update the tree accordingly

### LINKED ISSUE #7112: [OTA Server] change "location" parameter code once CHAR_STRING is supported
#### Problem
`emberAfOtaSoftwareUpdateServerClusterQueryImageCallback()` contains some bandaid code to get the length of the `location` parameter, since it is being generated as `uint8_t*` instead of `Span`:
```
    const uint8_t locationLen = emberAfStringLength(location);
    if ((locationLen != kLocationParamLength) || (metadataForServer.size() > kMaxMetadataLen))
```
(This was added in #6476)

This may need to happen in the same PR as when `CHAR_STRING` is supported, since it may not compile afterwards.

 #### Proposed Solution
After CHAR_STRING is supported, `location` should be a `Span` object, and the `size()` method can be used instead.

### PR DESCRIPTION
**Pull request by vivien-apple**
#### Problem

This PR uses `chip::ByteSpan` for `OCTET_STRING`, `CHAR_STRING`, `LONG_OCTET_STRING` and `LONG_CHAR_STRING`.

I'm keeping it as a draft at the moment since there are various codepaths that I have not tested locally...

#fixes  #7112, #7655, #6112

#### Change overview
 * Use `chip::ByteSpan` under the hood for `OCTET_STRING`, `CHAR_STRING`, `LONG_OCTET_STRING` and `LONG_CHAR_STRING`.
 * Update XML definitions for structs that were using `OCTET_STRING` instead of `CHAR_STRING`
 * Update various callbacks that were expecting a `uint8_t *`

#### Testing
 * It was tested locally but I still need to do various checks before removing it from the draft state. Particularly because I have updated the code of some clusters that does not have any tests...

### CODE REVIEW

File: src/app/clusters/groups-server/groups-server.cpp [hunk 1]
```diff
@@ -218,7 +219,7 @@ bool emberAfGroupsClusterViewGroupCallback(chip::app::Command * commandObj, Grou
         VerifyOrExit((writer = commandObj->GetCommandDataElementTLVWriter()) != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
         SuccessOrExit(err = writer->Put(TLV::ContextTag(0), status));
         SuccessOrExit(err = writer->Put(TLV::ContextTag(1), groupId));
-        SuccessOrExit(err = writer->PutString(TLV::ContextTag(2), reinterpret_cast<char *>(&groupName[1]), groupName[0]));
+        SuccessOrExit(err = writer->Put(TLV::ContextTag(2), groupName));
```
Comment (bzbarsky-apple) [inline src/app/clusters/groups-server/groups-server.cpp:222]: This is changing the encoding from string to octet string, in terms of the actual type in the TLV and hence what the other side will interpret this as.  Is this intended?  

If we're going to represent strings as `ByteSpan` (which I am stll not convinced about; `Span<char>` or the like would make way more sense to me), we should have a `PutString` overload that takes `ByteSpan`.

File: src/app/clusters/groups-server/groups-server.cpp [hunk 2]
```diff
@@ -389,8 +390,9 @@ bool emberAfGroupsClusterRemoveAllGroupsCallback(chip::app::Command * commandObj
                 }
                 else
                 {
-                    uint8_t groupName[ZCL_GROUPS_CLUSTER_MAXIMUM_NAME_LENGTH + 1] = { 0 };
-                    GroupId groupId                                               = binding.groupId;
+                    GroupId groupId                                                 = binding.groupId;
+                    uint8_t groupNameBuffer[ZCL_GROUPS_CLUSTER_MAXIMUM_NAME_LENGTH] = { 0 };
+                    chip::ByteSpan groupName(groupNameBuffer, sizeof(groupNameBuffer));
```
Comment (bzbarsky-apple) [inline src/app/clusters/groups-server/groups-server.cpp:395]:
```suggestion
                    chip::ByteSpan groupName(groupNameBuffer);
```
And it will infer the sizeof() bit from the type.

File: src/app/clusters/door-lock-server/door-lock-server-user.cpp [hunk 3]
```diff
@@ -725,12 +730,12 @@ bool emberAfDoorLockClusterUnlockDoorCallback(chip::app::Command * commandObj, u
                                (uint8_t *) &rfOperationEventMask, sizeof(rfOperationEventMask));
 
     // send operation event
-    if (doorUnlocked && (rfOperationEventMask & EMBER_BIT(2)) && (pin != NULL))
+    if (doorUnlocked && (rfOperationEventMask & EMBER_BIT(2)) && (pin.size() != 0))
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server-user.cpp:733]:
```suggestion
    if (doorUnlocked && (rfOperationEventMask & EMBER_BIT(2)) && !pin.empty())
```
And same in a bunch of other places.

**Comment by lazarkov** [comment]
@vivien-apple @bzbarsky-apple Will this PR be re-opened?

**Comment by vivien-apple** [comment]
I just rebased this PR. At the moment the code is still using `chip::ByteSpan`. I kept it for now because it is convenient and this PR is already changing a lot of files.
I would expect that converting from `ByteSpan` to `Span<char>` would be straightforward once everything is a `Span`, and 

If possible I would prefer to do it into a separate PR. The main reason for that is because the `uint8_t *` is creeping into more places and the use of `OCTET_STRING` instead of `CHAR_STRING` keeps creeping too.
This PR will at least lets people uses a `Span` derived class instead of `uint8_t *` and it will also fix the various places where `OCTET_STRING` is used instead of `CHAR_STRING`.

**Comment by bzbarsky-apple** [comment]
> I would expect that converting from ByteSpan to Span<char> would be straightforward once everything is a Span

Sounds like a plan.

File: src/app/clusters/account-login-server/account-login-server.cpp [hunk 4]
```diff
@@ -50,21 +50,21 @@ void sendResponse(app::CommandHandler * command, const char * responseSetupPin)
...
-                                             uint8_t * tempSetupPin)
+bool emberAfAccountLoginClusterLoginCallback(EndpointId endpoint, app::CommandHandler * command,
+                                             chip::ByteSpan tempAccountIdentifier, chip::ByteSpan tempSetupPin)
 {
-    // TODO: char is not null terminated, verify this code once #7963 gets merged.
-    std::string tempAccountIdentifierString(reinterpret_cast<char *>(tempAccountIdentifier));
-    std::string tempSetupPinString(reinterpret_cast<char *>(tempSetupPin));
+    std::string tempAccountIdentifierString(reinterpret_cast<char const *>(tempAccountIdentifier.data()),
+                                            tempAccountIdentifier.size());
+    std::string tempSetupPinString(reinterpret_cast<char const *>(tempSetupPin.data()), tempSetupPin.size());
```
Comment (lazarkov) [inline src/app/clusters/account-login-server/account-login-server.cpp:67]: Great to see this code uncommented.

**Comment by vivien-apple** [comment]
There are some conflicts but I guess this PR can be reviewed as if and I will update it when needed.

**Comment by vivien-apple** [comment]
@lazarkov @bzbarsky-apple @andreilitvin ?

**Review by woody-apple — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

**Review by lazarkov — APPROVED** [review]

File: src/app/clusters/door-lock-server/door-lock-server-user.cpp [hunk 5]
```diff
@@ -158,7 +158,7 @@ static void printUserTables(void)
 }
 
 // Returns status byte for use in SetPinResponse and SetRfidResponse commands.
-static uint8_t setUser(uint16_t userId, uint8_t userStatus, uint8_t userType, uint8_t * code,
+static uint8_t setUser(uint16_t userId, uint8_t userStatus, uint8_t userType, chip::ByteSpan code,
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server-user.cpp:161]:
```suggestion
static uint8_t setUser(uint16_t userId, uint8_t userStatus, uint8_t userType, ByteSpan code,
```
and same elsewhere in this file...

File: src/app/clusters/door-lock-server/door-lock-server-user.cpp [hunk 6]
```diff
@@ -620,7 +623,7 @@ static bool verifyPin(uint8_t * pin, uint8_t * userId)
     return false;
 }
 
-bool emberAfDoorLockClusterLockDoorCallback(EndpointId endpoint, app::CommandHandler * commandObj, uint8_t * PIN)
+bool emberAfDoorLockClusterLockDoorCallback(EndpointId endpoint, app::CommandHandler * commandObj, chip::ByteSpan PIN)
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server-user.cpp:626]: Why is PIN even CHAR_STRING?  It's octstr in the spec...  Followup issue to fix that?

File: src/app/clusters/general-commissioning-server/general-commissioning-server.cpp [hunk 7]
```diff
@@ -47,11 +52,15 @@ bool emberAfGeneralCommissioningClusterCommissioningCompleteCallback(EndpointId
...
 
 bool emberAfGeneralCommissioningClusterSetRegulatoryConfigCallback(EndpointId endpoint, app::CommandHandler * commandObj,
-                                                                   uint8_t location, uint8_t * countryCode, uint64_t breadcrumb,
-                                                                   uint32_t timeoutMs)
+                                                                   uint8_t location, chip::ByteSpan countryCode,
+                                                                   uint64_t breadcrumb, uint32_t timeoutMs)
 {
+    char countryCodePtr[kMaxCountryCodeLen + 1];
+    Platform::CopyString(countryCodePtr, sizeof(countryCodePtr), countryCode);
+    countryCodePtr[kMaxCountryCodeLen] = '\0';
```
Comment (bzbarsky-apple) [inline src/app/clusters/general-commissioning-server/general-commissioning-server.cpp:60]: Why is this needed?  CopyString ensures null-termination.

File: src/app/clusters/general-commissioning-server/general-commissioning-server.cpp [hunk 8]
```diff
@@ -47,11 +52,15 @@ bool emberAfGeneralCommissioningClusterCommissioningCompleteCallback(EndpointId
...
+
     CHIP_ERROR err = DeviceLayer::Internal::DeviceControlServer::DeviceControlSvr().SetRegulatoryConfig(
-        location, reinterpret_cast<const char *>(countryCode), breadcrumb);
+        location, reinterpret_cast<const char *>(countryCodePtr), breadcrumb);
```
Comment (bzbarsky-apple) [inline src/app/clusters/general-commissioning-server/general-commissioning-server.cpp:63]: This cast is not needed.  But also, followup issue to change SetRegulatoryConfig to not require a null-terminated string, please.

File: src/app/clusters/groups-server/groups-server.cpp [hunk 9]
```diff
@@ -89,7 +89,7 @@ void emberAfGroupsClusterServerInitCallback(EndpointId endpoint)
 // The first two bytes of the identifier is set to the groupId
 // The local endpoint is set to the endpoint that is mapped to this group
 // --------------------------
-static EmberAfStatus addEntryToGroupTable(EndpointId endpoint, GroupId groupId, uint8_t * groupName)
+static EmberAfStatus addEntryToGroupTable(EndpointId endpoint, GroupId groupId, chip::ByteSpan groupName)
```
Comment (bzbarsky-apple) [inline src/app/clusters/groups-server/groups-server.cpp:92]: Again, please lose the unnecessary chip:: prefixes.

File: src/app/clusters/groups-server/groups-server.cpp [hunk 10]
```diff
@@ -138,7 +138,8 @@ static EmberAfStatus removeEntryFromGroupTable(EndpointId endpoint, GroupId grou
         EmberStatus status   = emberDeleteBinding(bindingIndex);
         if (status == EMBER_SUCCESS)
         {
-            uint8_t groupName[ZCL_GROUPS_CLUSTER_MAXIMUM_NAME_LENGTH + 1] = { 0 };
+            uint8_t groupNameBuffer[ZCL_GROUPS_CLUSTER_MAXIMUM_NAME_LENGTH] = { 0 };
+            chip::ByteSpan groupName(groupNameBuffer, sizeof(groupNameBuffer));
```
Comment (bzbarsky-apple) [inline src/app/clusters/groups-server/groups-server.cpp:142]: This does not make sense to me.  This _used_ to pass in an empty string (first byte was 0).  Now it's passing in a span of size `ZCL_GROUPS_CLUSTER_MAXIMUM_NAME_LENGTH`.  This should just pass an empty (default-constructed) span, no?

File: src/app/clusters/groups-server/groups-server.cpp [hunk 11]
```diff
@@ -390,8 +391,9 @@ bool emberAfGroupsClusterRemoveAllGroupsCallback(EndpointId aEndpoint, app::Comm
                 }
                 else
                 {
-                    uint8_t groupName[ZCL_GROUPS_CLUSTER_MAXIMUM_NAME_LENGTH + 1] = { 0 };
-                    GroupId groupId                                               = binding.groupId;
+                    GroupId groupId                                                 = binding.groupId;
+                    uint8_t groupNameBuffer[ZCL_GROUPS_CLUSTER_MAXIMUM_NAME_LENGTH] = { 0 };
+                    chip::ByteSpan groupName(groupNameBuffer);
```
Comment (bzbarsky-apple) [inline src/app/clusters/groups-server/groups-server.cpp:396]: Again, this does not make sense to me.  We used to pass in an empty name, and now we are passing in a decidedly non-empty one, but full of null chars.

File: src/app/clusters/groups-server/groups-server.cpp [hunk 12]
```diff
@@ -190,12 +191,12 @@ bool emberAfGroupsClusterAddGroupCallback(EndpointId endpoint, app::CommandHandl
...
-    EmberAfStatus status                                          = EMBER_ZCL_STATUS_NOT_FOUND;
-    CHIP_ERROR err                                                = CHIP_NO_ERROR;
-    uint8_t groupName[ZCL_GROUPS_CLUSTER_MAXIMUM_NAME_LENGTH + 1] = { 0 };
+    EmberAfStatus status = EMBER_ZCL_STATUS_NOT_FOUND;
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+    chip::ByteSpan groupName;
 
     // Get the group name, if supported
-    emberAfPluginGroupsServerGetGroupNameCallback(emberAfCurrentEndpoint(), groupId, groupName);
+    emberAfPluginGroupsServerGetGroupNameCallback(emberAfCurrentEndpoint(), groupId, &groupName);
```
Comment (bzbarsky-apple) [inline src/app/clusters/groups-server/groups-server.cpp:199]: This also does not make sense.  What would make sense here would be passing in a buffer with `ZCL_GROUPS_CLUSTER_MAXIMUM_NAME_LENGTH` space in it and having the callee fill it in and adjust the size as needed.

File: src/app/clusters/media-input-server/media-input-server.cpp [hunk 13]
```diff
@@ -74,10 +74,10 @@ bool emberAfMediaInputClusterHideInputStatusCallback(EndpointId endpoint, app::C
     return true;
 }
 
-bool emberAfMediaInputClusterRenameInputCallback(EndpointId endpoint, app::CommandHandler * command, uint8_t input, uint8_t * name)
+bool emberAfMediaInputClusterRenameInputCallback(EndpointId endpoint, app::CommandHandler * command, uint8_t input,
+                                                 chip::ByteSpan name)
```
Comment (bzbarsky-apple) [inline src/app/clusters/media-input-server/media-input-server.cpp:78]: Again, extra chip:: should go.  This applies across the board here.

File: src/app/clusters/scenes/scenes.cpp [hunk 14]
```diff
@@ -718,23 +718,22 @@ EmberAfStatus emberAfScenesClusterRecallSavedSceneCallback(EndpointId endpoint,
...
                                             uint8_t * extensionFieldSets)
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
     EmberAfSceneTableEntry entry;
     EmberAfStatus status;
     bool enhanced                  = (cmd->commandId == ZCL_ENHANCED_ADD_SCENE_COMMAND_ID);
     uint16_t extensionFieldSetsLen = static_cast<uint16_t>(
-        cmd->bufLen -
-        (cmd->payloadStartIndex + sizeof(groupId) + sizeof(sceneId) + sizeof(transitionTime) + emberAfStringLength(sceneName) + 1));
+        cmd->bufLen - (cmd->payloadStartIndex + sizeof(groupId) + sizeof(sceneId) + sizeof(transitionTime) + sceneName.size() + 1));
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes/scenes.cpp:729]: This all does not make sense with TLV encoding.  Please file a followup issue on this?

File: src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt [hunk 15]
```diff
@@ -842,8 +842,7 @@ bool emberAf{{asUpperCamelCase parent.name}}Cluster{{asUpperCamelCase name}}Call
     {{#if (isOctetString type)}}
     ChipLogProgress(Zcl, "  {{asSymbol label}}: %zu", {{asSymbol label}}.size());
     {{else if (isCharString type)}}
-    // Currently the generated code emits `uint8_t *` for CHAR_STRING, it needs to emits ByteSpan
-    // ChipLogProgress(Zcl, "  {{asSymbol label}}: %.*s", {{asSymbol label}}.size(), {{asSymbol label}}.data());
+    ChipLogProgress(Zcl, " {{asSymbol label}}: %.*s", static_cast<uint16_t>({{asSymbol label}}.size()), {{asSymbol label}}.data());
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt:845]:
```suggestion
    ChipLogProgress(Zcl, " {{asSymbol label}}: %.*s", {{asSymbol label}}.size(), {{asSymbol label}}.data());
```

File: src/controller/java/templates/CHIPClusters-JNI.zapt [hunk 16]
```diff
@@ -276,8 +276,8 @@ class CHIP{{chipCallback.name}}AttributeCallback : public Callback::Callback<{{c
             err = JniReferences::GetInstance().FindMethod(env, javaCallbackRef, "onSuccess", "(Ljava/lang/String;)V", &javaMethod);
             VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(Zcl, "Could not find onSuccess method"));
 
-            UtfString valueStr(env, value);
-            env->CallVoidMethod(javaCallbackRef, javaMethod, valueStr.jniValue());
+            jstring valueStr = env->NewString(reinterpret_cast<const jchar *>(value.data()), value.size());
```
Comment (bzbarsky-apple) [inline src/controller/java/templates/CHIPClusters-JNI.zapt:279]: This doesn't seem right.  `jchar` is a 16-bit type, no?  This is going to run off the end of the buffer, and produce garbage for the part before that....  Seems to me like continuing to use UtfString but with the span would make more sense....

File: src/controller/java/templates/CHIPClusters-JNI.zapt [hunk 17]
```diff
@@ -350,6 +349,10 @@ class CHIP{{asUpperCamelCase parent.name}}Cluster{{asUpperCamelCase name}}Callba
             env->ExceptionClear();
             env->SetByteArrayRegion({{asSymbol label}}Arr, 0, {{asSymbol label}}.size(), reinterpret_cast<const jbyte *>({{asSymbol label}}.data()));
             VerifyOrExit(!env->ExceptionCheck(), err = CHIP_JNI_ERROR_EXCEPTION_THROWN);
+            {{else if (isCharString type)}}
+            {{asSymbol label}}Str = env->NewString(reinterpret_cast<const jchar *>({{asSymbol label}}.data()), {{asSymbol label}}.size());
```
Comment (bzbarsky-apple) [inline src/controller/java/templates/CHIPClusters-JNI.zapt:353]: Again, casting UTF-8 chars to `jchar*` does not make sense.

File: src/controller/java/templates/CHIPClusters-JNI.zapt [hunk 18]
```diff
@@ -458,7 +463,7 @@ class CHIP{{asUpperCamelCase parent.name}}{{asUpperCamelCase name}}AttributeCall
                 jbyteArray {{asLowerCamelCase name}} = env->NewByteArray(entries[i].{{name}}.size());
                 env->SetByteArrayRegion({{asLowerCamelCase name}}, 0, entries[i].{{name}}.size(), reinterpret_cast<const jbyte *>(entries[i].{{name}}.data()));
                 {{else if (isCharString type)}}
-                // Implement after ByteSpan is emitted instead of uint8_t *.
+                jstring {{asLowerCamelCase name}} = env->NewString(reinterpret_cast<const jchar *>(entries[i].{{name}}.data()), entries[i].{{name}}.size());
```
Comment (bzbarsky-apple) [inline src/controller/java/templates/CHIPClusters-JNI.zapt:466]: Can't cast to `jchar *`.

File: src/controller/java/templates/CHIPClusters-JNI.zapt [hunk 19]
```diff
@@ -477,7 +482,7 @@ class CHIP{{asUpperCamelCase parent.name}}{{asUpperCamelCase name}}AttributeCall
                 jbyteArray {{asLowerCamelCase name}} = env->NewByteArray(entries[i].size());
                 env->SetByteArrayRegion({{asLowerCamelCase name}}, 0, entries[i].size(), reinterpret_cast<const jbyte *>(entries[i].data()));
                 {{else if (isCharString type)}}
-                // Implement after ByteSpan is emitted instead of uint8_t *
+                jstring {{asLowerCamelCase name}} = env->NewString(reinterpret_cast<const jchar *>(entries[i].{{name}}.data()), entries[i].{{name}}.size());
```
Comment (bzbarsky-apple) [inline src/controller/java/templates/CHIPClusters-JNI.zapt:485]: Can't cast to `jchar *`.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/clusters/door-lock-server/door-lock-server-user.cpp (hunk 6 above)
Comment (lazarkov) [reply src/app/clusters/door-lock-server/door-lock-server-user.cpp:626]: I think octstr was not supported at the time.
